drivers/TickerClock.h

```diff
@@ -0,0 +1,87 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2019 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_TICKERCLOCK_H
+#define MBED_TICKERCLOCK_H
+
+#include <chrono>
+#include "hal/ticker_api.h"
+
+namespace mbed {
+/**
+ * \defgroup drivers_TickerClock TickerClock class
+ * \ingroup drivers-public-api-ticker
+ * @{
+ */
+
+/**
+ * A partial implementation of a C++11 Clock representing a HAL ticker.
+ *
+ * This class allows us to create chrono time_points for objects like Timer,
+ * with the limitation that the tickers are not singletons. This means:
+ *
+ * * the now() function is not static - this will limit
+ *   use with some algorithms,
+ * * there is no distinction between time_points for different
+ *   tickers
+ *
+ * This "pseudo-Clock" approach has been endorsed by Howard Hinnant
+ * (designer of Chrono) here:
+ * https://stackoverflow.com/questions/56400313/why-does-the-c-standard-require-the-clocknow-function-to-be-static
+ *
+ * TickerClock::time_point values should only be used with mbed APIs specifically taking
+ * them, not passed to generic templated chrono algorithms, and it is up to the user to use them
+ * in conjunction with the correct TickerClock.
+ *
+ * operators for `->` and conversion to `ticker_data_t *` are provided allowing
+ * TickerClock to be easily substituted in place of a `ticker_data_t *`.
+ */
+struct TickerClock {
```

Comment:
I noticed you are using struct instead of class. I know it's ok, but are you following some guidelines/convention or just because?

Reply:
It has no semantic meaning at all, apart from the minor detail of what the default member/base visibility is. You can use it just for that, just to save typing `public:`.

But in my head, I tend to feel happy calling "non-owning" copyable things "structs". "Class" has a vague suggestion of a deeper structure.

I think I apply a similar thing for `class` versus `typename` in templates. "Class" usually hints at something deep - this is likely a fancy object type, "typename" suggests it's likely a primitive type.

But not sure that it's worth having a hard and fast rule here.

Reply:
I thought it's worth seeing others' view. The [C++ core guidelines](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rc-class) disagree and say 

> C.8: Use class rather than struct if any member is non-public
> Reason
> Readability. To make it clear that something is being hidden/abstracted. This is a useful convention.

Reply:
Fluent C++ has a good take on this as well:

https://www.fluentcpp.com/2017/06/13/the-real-difference-between-struct-class/

Quoting:

> The C++ Core Guidelines
> The above has been inspired by the C++ Core Guideline (which is a great read by the way), in particular the following:
> 
> C.1: Organize related data into structures (structs or classes)
> C.2: Use class if the class has an invariant; use struct if the data members can vary independently
> C.3: Represent the distinction between an interface and an implementation using a class
> C.8: Use class rather than struct if any member is non-public

I like the idea of C.2 and of struct equivalents to bundle, whereas a class can do things.

Also doing a lot of Swift development, structs are value types whereas classes are reference types.  Which tends to agree with C.2 --> use struct for things that can live independently.

I don't mind having a struct do things as well and not just store data.

Reply:
It's worth noting that in this particular case, there *is* a private member, but it's actually not really hidden, as there is an implicit conversion operator that gives it straight to you.

This really supposed to just be an enhanced `ticker_data *` that is usable as that, but also adds the `Clock` API.

A more simple clock is actually a bit odd anyway, as they usually have *no* members, and only typedefs and a static `now` function. So they technically don't hit C.8, but then they're not a "bundle" either, they're just an interface (to `now`).